cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/kafkaassigner/KafkaAssignerDiskUsageDistributionGoal.java

```diff
@@ -0,0 +1,568 @@
+              broker, diskUsage(broker), meanDiskUsage);
+    double brokerDiskUsage = diskUsage(broker);
+    boolean improved = false;
+    if (brokerDiskUsage > upperThreshold) {
+      List<Broker> brokersAscend = clusterModel.sortedHealthyBrokersUnderThreshold(DISK, brokerDiskUsage);
+      for (Broker toSwapWith : brokersAscend) {
+        if (toSwapWith == broker) {
+          continue;
+        }
+        if (swapReplicas(broker, toSwapWith, meanDiskUsage, clusterModel, excludedTopics)) {
+          improved = true;
+          break;
+        }
+      }
+    } else if (brokerDiskUsage < lowerThreshold) {
+      List<Broker> brokersDescend = sortedBrokersAboveSizeDescend(clusterModel, brokerDiskUsage);
+      for (Broker toSwapWith : brokersDescend) {
+        if (toSwapWith == broker) {
+          continue;
+        }
+        if (swapReplicas(broker, toSwapWith, meanDiskUsage, clusterModel, excludedTopics)) {
+          improved = true;
+          break;
+        }
+      }
+    }
+    return improved;
+  }
+
+  /**
+   * Swap replica between two brokers. The method should achieve the result that the overall usage of the two 
+   * brokers are improved. More specifically, the following result is reduced.
+   * <pre>
+   * (<tt>UsageOfBroker1</tt> - <tt>MeanUsage</tt>) + (<tt>UsageOfBroker2</tt> - <tt>MeanUsage</tt>)
+   * </pre>
+   * 
+   * @param toSwap the broker that needs to swap a replica with the other broker.
+   * @param toSwapWith the broker that provides a replica to swap with the broker <tt>toSwap</tt>
+   * @param meanDiskUsage the average usage of the cluster.
+   * @param clusterModel the cluster model.
+   * @return true if a swap has been done, false otherwise.
+   * 
+   * @throws AnalysisInputException
+   */
+  boolean swapReplicas(Broker toSwap, 
+                       Broker toSwapWith, 
+                       double meanDiskUsage, 
+                       ClusterModel clusterModel,
+                       Set<String> excludedTopics) throws AnalysisInputException {
+    LOG.trace("Swapping replicas between broker {}({}) and broker {}({})",
+             toSwap.id(), brokerSize(toSwap), toSwapWith.id(), brokerSize(toSwapWith));
+    double sizeToChange = toSwap.capacityFor(DISK) * meanDiskUsage - brokerSize(toSwap);
+    List<ReplicaWrapper> sortedReplicasToSwap = sortReplicasAscend(toSwap.replicas(), excludedTopics);
+    List<ReplicaWrapper> sortedLeadersToSwapWith = sortReplicasAscend(toSwapWith.leaderReplicas(), excludedTopics);
+    List<ReplicaWrapper> sortedFollowersToSwapWith = sortReplicasAscend(followerReplicas(toSwapWith), excludedTopics);
+    
+    int startPos;
+    int delta;
+    if (sizeToChange > 0) {
+      // iterate from small replicas to large replicas.
+      startPos = 0;
+      delta = 1;
+    } else {
+      // iterate from large replicas to small replicas.
+      startPos = sortedReplicasToSwap.size() - 1;
+      delta = -1;
+    }
```

Comment:
It seems we are emulating a deque functionality with a sorted list. Would it be possible to replace `sortedReplicasToSwap` with a deque and iterate with `descendingIterator` for large to small iteration?

Reply:
We can use Deque. I was using a List because `sortReplicasAscend` returns a List and I wanted to avoid memory copy (to create a Deque from a List).